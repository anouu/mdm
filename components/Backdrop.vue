<template>
  <div
    :class="['backdrop', { 'backdrop--open': isBackdropOpen }]"
    @click="closeAll()"
  ></div>
</template>

<script>
import { mapState, mapActions } from 'vuex'

export default {
  computed: {
    ...mapState('ui', ['isBackdropOpen']),
  },

  methods: {
    ...mapActions('ui', ['closeAll']),
  },
}
</script>

<style lang="scss" scoped>
.backdrop {
  position: fixed;
  inset: 0;
  background-color: rgba(#000, 0.5);
  opacity: 0;
  transform: translateX(100%);
  transition: opacity var(--transition-duration) $ease-in-sine
      var(--transition-duration),
    transform 0ms linear 300ms;
  will-change: transition, opacity;
  cursor: pointer;
  z-index: 50;

  &--open {
    transition-delay: 0ms;
    opacity: 1;
    transform: translateX(0);
  }
}
</style>
