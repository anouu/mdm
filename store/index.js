export const state = () => ({
  cart: {},
})

export const getters = {
  getCartItems: (state) => {
    return state.cart?.products || []
  },

  getCartQty: (state, getters) => {
    return Object.values(getters.getCartItems).reduce(
      (acc, item) => acc + item.qty,
      0
    )
  },

  getCartTotal: (state, getters) => {
    return Object.values(getters.getCartItems).reduce(
      (acc, item) => acc + item.qty * item.price.base.amount,
      0
    )
  },

  isCartEmpty: (state, getters) => {
    return getters.getCartQty === 0
  },
}

export const mutations = {
  SET_CART(state, newCart) {
    state.cart = newCart
  },

  UPDATE_PRODUCT(state, { reference, updatedItem }) {
    state.cart.products[reference] = updatedItem
  },
}

export const actions = {
  async setSampleCart({ commit }) {
    try {
      const { cart } = await fetch('/api/sampleCart').then((res) => res.json())

      commit('SET_CART', cart)
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error(e)
    }
  },

  updateProduct({ commit, getters }, { reference, qty }) {
    const items = getters.getCartItems
    const updatedItem = { ...items[reference], qty }

    commit('UPDATE_PRODUCT', { reference, updatedItem })
  },

  deleteProduct(context, reference) {
    const items = context.getters.getCartItems
    const newCart = { ...context.state.cart, products: items }

    delete items[reference]

    context.commit('SET_CART', newCart)
    context.commit('ui/SET_MODAL_OPEN', false)
  },
}
