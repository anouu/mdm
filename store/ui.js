export const state = () => ({
  isBackdropOpen: false,
  isModalOpen: false,
  isBasketMiniOpen: false,
  isDarkMode: false,
  modalData: {
    status: false,
    data: {
      body: '',
      action: null,
      actionData: null,
    },
  },
})

export const mutations = {
  SET_BACKDROP_OPEN(state, boolean) {
    state.isBackdropOpen = boolean
  },

  SET_MODAL_OPEN(state, boolean) {
    state.isModalOpen = boolean
    state.isBackdropOpen = boolean
  },

  SET_DRAWER_OPEN(state, boolean) {
    state.isBasketMiniOpen = boolean
    state.isBackdropOpen = boolean
  },

  CLOSE_ALL_OPEN(state) {
    state.isBackdropOpen = false
    state.isModalOpen = false
    state.isBasketMiniOpen = false
  },

  SET_MODAL_DATA(state, data) {
    state.modalData = data
  },

  SET_DARK_MODE(state, boolean) {
    state.isDarkMode = boolean
  },
}

export const actions = {
  closeAll({ commit }) {
    commit('CLOSE_ALL_OPEN', false)
  },

  setBasketMini({ commit }, boolean) {
    commit('SET_DRAWER_OPEN', boolean)
  },

  setModal({ commit }, { status, data }) {
    commit('SET_MODAL_DATA', data)
    commit('SET_MODAL_OPEN', status)
  },

  setDarkMode({ commit }, boolean) {
    commit('SET_DARK_MODE', boolean)
  },
}
