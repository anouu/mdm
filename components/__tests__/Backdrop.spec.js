import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import * as Ui from '../../store/ui'
import Backdrop from '../Backdrop'

const localVue = createLocalVue()

localVue.use(Vuex)

describe('Backdrop', () => {
  let store
  let factory

  beforeEach(() => {
    store = new Vuex.Store({
      modules: {
        ui: {
          state: Ui.state,
          actions: Ui.actions,
          namespaced: true,
        },
      },
    })

    factory = () => {
      return shallowMount(Backdrop, { store, localVue })
    }
  })

  test('mounts properly', () => {
    const wrapper = factory()
    expect(wrapper.vm).toBeTruthy()
  })

  test('renders properly', () => {
    const wrapper = factory()
    expect(wrapper.html()).toMatchSnapshot()
  })
})
