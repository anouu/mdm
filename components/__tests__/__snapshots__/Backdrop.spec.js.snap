// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backdrop renders properly 1`] = `<div class="backdrop"></div>`;
